src/Plutip/Types.purs

```diff
@@ -135,3 +168,34 @@ instance DecodeAeson StopClusterResponse where
         StopClusterFailure <$> decodeAeson failure
       _ -> do
         Left (UnexpectedValue (toStringifiedNumbersJson aeson))
+
+type InitialUtxos = Array UtxoAmount
+
+-- | A type class that implements a type-safe interface for specifying UTXO
+-- | distribution for wallets.
+-- | Number of wallets in distribution specification matches the number of
+-- | wallets provided to the user.
+class UtxoDistribution distr wallets | distr -> wallets, wallets -> distr where
+  encodeDistribution :: distr -> Array (Array UtxoAmount)
+  decodeWallets :: Array PrivateKeyResponse -> Maybe wallets
+
+instance UtxoDistribution Unit Unit where
+  encodeDistribution _ = []
+  decodeWallets _ = Just unit
+
+instance UtxoDistribution InitialUtxos KeyWallet where
+  encodeDistribution amounts = [ amounts ]
+  decodeWallets [ (PrivateKeyResponse key) ] =
+    pure $ privateKeysToKeyWallet (PrivatePaymentKey key) Nothing
+  decodeWallets _ = Nothing
+
+instance
+  UtxoDistribution restSpec restWallets =>
+  UtxoDistribution (InitialUtxos /\ restSpec) (KeyWallet /\ restWallets) where
+  encodeDistribution (amounts /\ rest) =
+    encodeDistribution amounts <> encodeDistribution rest
+  decodeWallets = Array.uncons >>> case _ of
+    Nothing -> Nothing
+    Just { head: PrivateKeyResponse key, tail } ->
+      Tuple (privateKeysToKeyWallet (PrivatePaymentKey key) Nothing) <$>
+        decodeWallets tail
```

Comment:
Instead of a lawless class, we could opt for a direct shallow embedding:

```haskell
newtype UtxoDistribution wallets
  = UtxoDistribution
      { distribution :: Array (Array UtxoAmount)
      , decodeWallets :: Array PrivateKeyResponse -> Maybe wallets
      }

nilDistribution =
  UtxoDistribution 
    { distribution: []
    , decodeWallets: const $ Just unit
    }

withWallet distribution (UtxoDistribution next) =
  UtxoDistribution
    { distribution: distribution : next.distribution
    , decodeWallets: Array.uncons >>> case _ of
        Nothing -> Nothing
        Just { head: PrivateKeyResponse key, tail } ->
          Tuple (privateKeysToKeyWallet (PrivatePaymentKey key) Nothing) <$>
            next.decodeWallets tail
    }

distribution
  = withWallet [UtxoAmount]
  $ withWallet [UtxoAmount, UtxoAmount]
    nilDistribution
```

Reply:
It's not lawless. The law is that the number of wallets in the spec matches the number of provided wallets.

Reply:
Oh I see, still, I'm not sure that's enough to make precise the expected behavior? For example, ordering in the `wallets` type